### PR TITLE
Add compiler change for deprecation date

### PR DESCRIPTION
**Pull request by yuli-han**
## Summary

Add change in compiler for deprecation date. 
1. Print out warning message when a group_by/join are flagged with deprecation date but there are downstream group_bys/joins with no deprecation date or later deprecation date. 
2. Print out warning message when a group_by/join are not flagged with deprecation date but there are upstream group_bys/joins with deprecation_date settings. 

The warning message will only be printed out for the configs user provided so there will be no duplicate message print out when we materializing downstream entities. 

## Test Plan

Integration tests: 
1. Test compile a join with to-be-deprecated groupBy
image
2. Test compile a groupBy with deprecation date while the downstream are not flagged deprecation_date
image
3. Test compile a groupBy with a to-be-deprecated join source
image
4. Test compile a to-be-deprecated join which is used as source in groupBy
image

- [x] Added Unit Tests
- [x] Covered by existing CI
- [x] Integration tested

## Checklist
- [ ] Documentation update

## Reviewers
@rohitgirme @pengyu-hou @donghanz @pkundurthy

### CODE REVIEW

File: api/py/ai/chronon/repo/compile.py [hunk 1]
```diff
@@ -163,13 +163,16 @@ def extract_and_convert(chronon_root, input_path, output_root, debug, force_over
...
             )
 
+            _handle_deprecation_warning(obj, obj_class, new_group_bys, new_joins)
+
             # Update the accumulated dictionaries with the new entries
             extra_dependent_group_bys_to_materialize.update(new_group_bys)
             extra_dependent_joins_to_materialize.update(new_joins)
 
+    dependencies = {}
+    dependencies.update({**extra_dependent_group_bys_to_materialize, **extra_dependent_joins_to_materialize})
```
Comment (rohitgirme) [inline api/py/ai/chronon/repo/compile.py:173]: Seems like the `dependencies` variable is only used in the IF block below (L175-L180). What's the reason to remove this out of the IF block?

File: api/py/test/sample/joins/sample_team/sample_deprecation_join.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+from ai.chronon.join import Join, JoinPart
+from group_bys.sample_team import sample_deprecation_group_by
+from sources import test_sources
+
+v1 = Join(
+    left=test_sources.staging_entities,
+    right_parts=[JoinPart(group_by=sample_deprecation_group_by.v1, tags={"experimental": True})],
+    table_properties={"config_json": """{"sample_key": "sample_value"}"""},
+    output_namespace="sample_namespace",
```
Comment (rohitgirme) [inline api/py/test/sample/joins/sample_team/sample_deprecation_join.py:23]: Suggest keeping example simple by removing attributes not required for test like `env, tags` etc

File: api/py/test/test_compile.py [hunk 3]
```diff
@@ -79,11 +79,37 @@ def test_basic_compile():
         extract_and_convert, ["--chronon_root=test/sample", "--input_path=joins/sample_team/sample_join.py"]
     )
     assert result.exit_code == 0
+
+
+def test_compile_deprecation():
```
Comment (rohitgirme) [inline api/py/test/test_compile.py:84]: Highly recommend keeping the 2 test cases in separate test methods. So each test is testing only 1 use case. Easier to debug if test fails

File: api/py/ai/chronon/repo/compile.py (hunk 1 above)
Comment (yuli-han) [reply api/py/ai/chronon/repo/compile.py]: updated

File: api/py/test/test_compile.py (hunk 3 above)
Comment (yuli-han) [reply api/py/test/test_compile.py]: updated

File: api/py/test/sample/joins/sample_team/sample_deprecation_join.py (hunk 2 above)
Comment (yuli-han) [reply api/py/test/sample/joins/sample_team/sample_deprecation_join.py:22]: Updated

File: api/py/ai/chronon/repo/compile.py [hunk 4]
```diff
@@ -286,6 +288,58 @@ def _print_features(obj: Union[Join, GroupBy], obj_class: Type[Union[Join, Group
...
+            for source in obj.sources:
+                if source.joinSource and source.joinSource.join.metaData.deprecationDate:
+                    _print_error(
+                        "Deprecation Warning:",
+                        f"Join source {source.joinSource.join.metaData.name} is going to be"
+                        f" deprecated by {source.joinSource.join.metaData.deprecationDate}. Please"
+                        f" ensure either to remove/migrate the dependency or set up correct deprecationDate"
+                        f" for {obj.metaData.name}.",
+                    )
+
```
Comment (pengyu-hou) [inline api/py/ai/chronon/repo/compile.py:341]: I used GPT to reorganize the code a bit: 🤣 This logic is hard to get right. 
```
def check_deprecation_date(collection, message):
    incorrect_deprecation_date = [
        name for name, item in collection.items()
        if not item.metaData.deprecationDate or item.metaData.deprecationDate > obj.metaData.deprecationDate
    ]

    if incorrect_deprecation_date:
        _print_error(
            "Deprecation Warning:",
            f"Downstream {message} {incorrect_deprecation_date}"
            f" are missing deprecationDate or has a later deprecationDate than upstream {obj.metaData.name}."
            f" Please ensure either to remove/migrate the dependency or set up correct deprecationDate."
        )

def check_deprecation_existence(collection, message):
    for name, item in collection.items():
        if item.metaData.deprecationDate:
            _print_error(
                "Deprecation Warning:",
                f"{message} {name} is going to be deprecated by"
                f" {item.metaData.deprecationDate}. Please ensure either to remove/migrate"
                f" the dependency or set up correct deprecationDate for {obj.metaData.name}.",
            )

if obj.metaData.deprecationDate:
    check_deprecation_date(downstream_group_bys, "group_bys")
    check_deprecation_date(downstream_joins, "joins")
else:
    if obj_class is Join:
        check_deprecation_existence({jp.groupBy.metaData.name: jp.groupBy for jp in obj.joinParts}, "Join part")
    else:
        check_deprecation_existence(
            {source.joinSource.join.metaData.name: source.joinSource.join 
            for source in obj.sources 
            if source.joinSource}, 
            "Join source"
        )
```

**Review by rohitgirme — APPROVED** [review]

File: api/py/ai/chronon/repo/compile.py (hunk 4 above)
Comment (yuli-han) [reply api/py/ai/chronon/repo/compile.py:374]: Updated

**Review by pengyu-hou — APPROVED** [review]

**Pull request merged**